File: src/tachikoma/workflows/loader.py
"""Workflow definition loader for discovering workflows from skill directories.

This module provides functionality to discover and parse workflow definitions
from the filesystem. Workflows are discovered as subdirectories within skill's
workflows/ directories, with steps parsed as subdirectories containing
instructions.md files with YAML frontmatter.
"""

from pathlib import Path

import frontmatter
from loguru import logger

from tachikoma.workflows.definition import StepDefinition, WorkflowDefinition

_log = logger.bind(component="workflows")


def load_workflows(skill_dir: Path, skill_name: str) -> dict[str, WorkflowDefinition]:
    """Discover and load workflow definitions from a skill directory.

    Scans the workflows/ subdirectory within skill_dir for workflow definitions.
    Each workflow is a subdirectory containing step subdirectories sorted
    alphabetically. Steps are parsed from instructions.md files with YAML frontmatter.

    Args:
        skill_dir: Path to the skill directory.
        skill_name: Name of the skill (used in workflow definitions).

    Returns:
        Dictionary mapping workflow name to WorkflowDefinition.
    """
    workflows_dir = skill_dir / "workflows"

    if not workflows_dir.exists() or not workflows_dir.is_dir():
        _log.debug(
            "No workflows directory found: skill={skill}, path={path}",
            skill=skill_name,
            path=str(workflows_dir),
        )
        return {}

    workflows: dict[str, WorkflowDefinition] = {}

    try:
        items = list(workflows_dir.iterdir())
    except Exception as exc:
        _log.warning(
            "Failed to list workflows directory: skill={skill}, path={path}, err={err}",
            skill=skill_name,
            path=str(workflows_dir),
            err=str(exc),
        )
        return {}

    for item in items:
        if not item.is_dir():
            continue

        workflow_name = item.name
        workflow_def = _load_workflow(item, skill_name, workflow_name)

        if workflow_def is not None:
            workflows[workflow_name] = workflow_def

    return workflows


def _load_workflow(
    workflow_dir: Path,
    skill_name: str,
    workflow_name: str,
) -> WorkflowDefinition | None:
    """Load a single workflow definition from its directory.

    Args:
        workflow_dir: Path to the workflow directory.
        skill_name: Name of the parent skill.
        workflow_name: Name of the workflow.

    Returns:
        WorkflowDefinition if at least one valid step found, None otherwise.
    """
    try:
        items = list(workflow_dir.iterdir())
    except Exception as exc:
        _log.warning(
            "Failed to list workflow directory: skill={skill}, workflow={workflow}, err={err}",
            skill=skill_name,
            workflow=workflow_name,
            err=str(exc),
        )
        return None

    step_dirs = sorted(
        [item for item in items if item.is_dir()],
        key=lambda p: p.name,
    )

    steps: list[StepDefinition] = []

    for step_dir in step_dirs:
        step_def = _load_step(step_dir, skill_name, workflow_name)

        if step_def is not None:
            steps.append(step_def)

    workflow_def = WorkflowDefinition(
        skill_name=skill_name,
        workflow_name=workflow_name,
        steps=steps,
        path=workflow_dir,
    )

    _log.debug(
        "Loaded workflow: skill={skill}, workflow={workflow}, steps={count}",
        skill=skill_name,
        workflow=workflow_name,
        count=len(steps),
    )

    return workflow_def


def _load_step(
    step_dir: Path,
    skill_name: str,
    workflow_name: str,
) -> StepDefinition | None:
    """Load a single step definition from its directory.

    Args:
        step_dir: Path to the step directory.
        skill_name: Name of the parent skill.
        workflow_name: Name of the parent workflow.

    Returns:
        StepDefinition if valid, None if instructions.md missing or invalid.
    """
    instructions_path = step_dir / "instructions.md"

    if not instructions_path.exists():
        _log.warning(
            "Step missing instructions.md: skill={skill}, workflow={workflow}, step={step}",
            skill=skill_name,
            workflow=workflow_name,
            step=step_dir.name,
        )
        return None

    try:
        post = frontmatter.load(str(instructions_path))
    except Exception as exc:
        _log.warning(
            "Failed to parse instructions.md: skill={skill}, workflow={workflow}, "
            "step={step}, err={err}",
            skill=skill_name,
            workflow=workflow_name,
            step=step_dir.name,
            err=str(exc),
        )
        return None

    title = post.metadata.get("title", "")

    if not title or not isinstance(title, str):
        _log.warning(
            "Step has missing or invalid title: skill={skill}, workflow={workflow}, step={step}",
            skill=skill_name,
            workflow=workflow_name,
            step=step_dir.name,
        )
        return None

    skippable = post.metadata.get("skippable", False)

    if not isinstance(skippable, bool):
        _log.warning(
            "Step has invalid skippable type: skill={skill}, workflow={workflow}, step={step}",
            skill=skill_name,
            workflow=workflow_name,
            step=step_dir.name,
        )
        return None

    properties = {k: v for k, v in post.metadata.items() if k not in ("title", "skippable")}

    references_path = step_dir / "references"
    if not references_path.exists() or not references_path.is_dir():
        references_path = None

    scripts_path = step_dir / "scripts"
    if not scripts_path.exists() or not scripts_path.is_dir():
        scripts_path = None

    step_def = StepDefinition(
        id=step_dir.name,
        title=title,
        instructions_path=instructions_path,
        references_path=references_path,
        scripts_path=scripts_path,
        skippable=skippable,
        properties=properties,
    )

    _log.debug(
        "Loaded step: skill={skill}, workflow={workflow}, step={step}, title={title}",
        skill=skill_name,
        workflow=workflow_name,
        step=step_dir.name,
        title=title,
    )

    return step_def

File: src/extensions/workflows/loader.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

import { parseFrontmatter } from "@earendil-works/pi-coding-agent";

import type { Logger } from "../../log.ts";
import { refKey } from "./composition.ts";

export interface StepDefinition {
  /** Step identifier (matches the directory name). */
  id: string;
  title: string;
  instructionsPath: string;
  referencesPath: string | null;
  scriptsPath: string | null;
  /** When false, the step can be skipped. */
  required: boolean;
  /** Natural-language predicate; when set the cascade halts so the agent decides start/skip. */
  condition: string | null;
  /** Reference to a sub-workflow run to completion when the step activates. */
  composes: string | null;
  /** Reference to a sub-workflow iterated once per agent-supplied item. */
  loop: string | null;
  /** Extensible frontmatter fields beyond the reserved keys. */
  properties: Record<string, unknown>;
}

export interface WorkflowDefinition {
  skillName: string;
  workflowName: string;
  /** Ordered by step directory name (e.g. 01-plan, 02-execute). */
  steps: StepDefinition[];
  path: string;
}

const RESERVED_STEP_KEYS = ["title", "required", "skippable", "condition", "composes", "loop"];

/**
 * Read an optional string-typed frontmatter field, warning and falling back to
 * null when present but mistyped — a bad value never rejects the whole step.
 */
const optionalString = (
  frontmatter: Record<string, unknown>,
  key: string,
  context: { skillName: string; workflowName: string; step: string },
  log: Logger,
): string | null => {
  const value = frontmatter[key];

  if (value == null) return null;

  if (typeof value !== "string") {
    log.warn({ ...context }, `step has invalid ${key} type (expected string) — treated as unset`);
    return null;
  }

  return value;
};

const isDirectory = (path: string): boolean => {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
};

const listDirectories = (path: string): string[] => {
  try {
    return readdirSync(path, { withFileTypes: true })
      .filter((entry) => entry.isDirectory())
      .map((entry) => entry.name);
  } catch {
    return [];
  }
};

const loadStep = (
  stepDir: string,
  stepId: string,
  context: { skillName: string; workflowName: string },
  log: Logger,
): StepDefinition | null => {
  const instructionsPath = join(stepDir, "instructions.md");

  if (!existsSync(instructionsPath)) {
    log.warn({ ...context, step: stepId }, "step missing instructions.md — skipped");
    return null;
  }

  let frontmatter: Record<string, unknown>;

  try {
    frontmatter = parseFrontmatter(readFileSync(instructionsPath, "utf8")).frontmatter;
  } catch (error) {
    log.warn({ err: error, ...context, step: stepId }, "failed to parse instructions.md — skipped");
    return null;
  }

  const title = frontmatter.title;

  if (typeof title !== "string" || title.length === 0) {
    log.warn({ ...context, step: stepId }, "step has missing or invalid title — skipped");
    return null;
  }

  let required = true;

  if ("required" in frontmatter) {
    if (typeof frontmatter.required !== "boolean") {
      log.warn({ ...context, step: stepId }, "step has invalid required type — skipped");
      return null;
    }

    required = frontmatter.required;
  } else if ("skippable" in frontmatter) {
    // skippable is the deprecated alias for required: false
    log.warn(
      { ...context, step: stepId },
      "step uses deprecated 'skippable' field; use 'required: false' instead",
    );
    required = frontmatter.skippable !== true;
  }

  const fieldContext = { ...context, step: stepId };
  const condition = optionalString(frontmatter, "condition", fieldContext, log);
  const composes = optionalString(frontmatter, "composes", fieldContext, log);
  const loop = optionalString(frontmatter, "loop", fieldContext, log);

  const properties = Object.fromEntries(
    Object.entries(frontmatter).filter(([key]) => !RESERVED_STEP_KEYS.includes(key)),
  );

  const referencesPath = join(stepDir, "references");
  const scriptsPath = join(stepDir, "scripts");

  return {
    id: stepId,
    title,
    instructionsPath,
    referencesPath: isDirectory(referencesPath) ? referencesPath : null,
    scriptsPath: isDirectory(scriptsPath) ? scriptsPath : null,
    required,
    condition,
    composes,
    loop,
    properties,
  };
};

const loadWorkflow = (
  workflowDir: string,
  skillName: string,
  workflowName: string,
  log: Logger,
): WorkflowDefinition => {
  const steps = listDirectories(workflowDir)
    .sort()
    .map((stepId) => loadStep(join(workflowDir, stepId), stepId, { skillName, workflowName }, log))
    .filter((step) => step != null);

  return { skillName, workflowName, steps, path: workflowDir };
};

/** Load every workflow defined under a skill directory's workflows/ subdirectory. */
export const loadSkillWorkflows = (
  skillDir: string,
  skillName: string,
  log: Logger,
): WorkflowDefinition[] => {
  const workflowsDir = join(skillDir, "workflows");

  return listDirectories(workflowsDir).map((workflowName) =>
    loadWorkflow(join(workflowsDir, workflowName), skillName, workflowName, log),
  );
};

/**
 * Load every workflow across every skill under the skills root, keyed by
 * `skill/workflow`. Used at bootstrap to validate the composition graph.
 */
export const loadAllWorkflows = (
  skillsRoot: string,
  log: Logger,
): Map<string, WorkflowDefinition> => {
  const all = new Map<string, WorkflowDefinition>();

  for (const skillName of listDirectories(skillsRoot)) {
    for (const def of loadSkillWorkflows(join(skillsRoot, skillName), skillName, log)) {
      all.set(refKey(skillName, def.workflowName), def);
    }
  }

  return all;
};

/**
 * Resolve a single workflow definition from the skills root, reading the
 * filesystem fresh on every call so definition edits apply without a restart.
 */
export const findWorkflow = (
  skillsRoot: string,
  skillName: string,
  workflowName: string,
  log: Logger,
): WorkflowDefinition | null => {
  const workflowDir = join(skillsRoot, skillName, "workflows", workflowName);

  if (!isDirectory(workflowDir)) return null;

  return loadWorkflow(workflowDir, skillName, workflowName, log);
};
